parsers/CA_NS.py

```diff
@@ -1,213 +1,198 @@
+    base_loads = session.get(LOAD_URL).json()  # Base loads in MW
+    mixes_percent = session.get(MIX_URL).json()  # Electricity breakdowns in %
+    if any(
+        base_load["datetime"] != mix_percent["datetime"]
+        for base_load, mix_percent in zip(base_loads, mixes_percent)
+    ):
+        raise ParserException(PARSER, "source data is out of sync", ZONE_KEY)
+
+    exchanges = ExchangeList(logger)
+    production_breakdowns = ProductionBreakdownList(logger)
+    # Skip the first element of each JSON array because the reported base load
+    # is always 0 MW.
+    for base_load, mix_percent in zip(base_loads[1:], mixes_percent[1:]):
+        # The datetime key is in the format '/Date(1493924400000)/'; extract
+        # the timestamp 1493924400 (cutting out the last three zeros as well).
+        date_time = datetime.fromtimestamp(
+            int(base_load["datetime"][6:-5]), tz=timezone.utc
+        )
 
-def _get_ns_info(requests_obj, logger: Logger):
-    zone_key = "CA-NS"
-
-    # This is based on validation logic in https://www.nspower.ca/site/renewables/assets/js/site.js
-    # In practical terms, I've seen hydro production go way too high (>70%) which is way more
-    # than reported capacity.
-    valid_percent = {
-        # The validation JS reports error when Solid Fuel (coal) is over 85%,
-        # but as far as I can tell, that can actually be a valid result, I've seen it a few times.
-        # Use 98% instead.
-        "coal": (0, 0.98),
-        "gas": (0, 0.5),
-        "biomass": (0, 0.15),
-        "hydro": (0, 0.60),
-        "wind": (0, 0.55),
-        "imports": (0, 0.50),
-    }
-
-    # Sanity checks: verify that reported production doesn't exceed listed capacity by a lot.
-    # In particular, we've seen error cases where hydro production ends up calculated as 900 MW
-    # which greatly exceeds known capacity of 418 MW.
-    valid_absolute = {
-        "coal": 1300,
-        "gas": 700,
-        "biomass": 100,
-        "hydro": 500,
-        "wind": 700,
-    }
-
-    mix_url = "https://www.nspower.ca/library/CurrentLoad/CurrentMix.json"
-    mix_data = requests_obj.get(mix_url).json()
-
-    load_url = "https://www.nspower.ca/library/CurrentLoad/CurrentLoad.json"
-    load_data = requests_obj.get(load_url).json()
-
-    production = []
-    imports = []
-    for mix in mix_data:
-        percent_mix = {
-            "coal": mix["Solid Fuel"] / 100.0,
-            "gas": (mix["HFO/Natural Gas"] + mix["CT's"] + mix["LM 6000's"]) / 100.0,
-            "biomass": mix["Biomass"] / 100.0,
-            "hydro": mix["Hydro"] / 100.0,
-            "wind": mix["Wind"] / 100.0,
-            "imports": mix["Imports"] / 100.0,
+        mix_fraction = {
+            "coal": mix_percent["Solid Fuel"] / 100.0,
+            "gas": (
+                mix_percent["HFO/Natural Gas"]
+                + mix_percent["CT's"]
+                + mix_percent["LM 6000's"]
+            )
+            / 100.0,
```

Comment:
I'm not sure what you mean. How would you split these? Which operations are unnecessary? This object maps the source's production modes (plus imports) to each of Electricity Maps' production modes (plus imports), but it only does that once for each entry in the source array.

Reply:
add_value() supports adding values for the same mode multiple times. So instead of doing additions or summing things in the parsers we should just be adding the source values directly.

This ensures that the parsers are kept as simple as possible and that all the source values are correlated on an individual basis if needed.

Reply:
Ahh, I think I see what you're saying. I'll try to use that instead, but it might be a bit tricky to integrate it properly because of the custom validation logic in this parser.